execution_engine/src/core/runtime/mod.rs

```diff
@@ -1337,38 +1352,51 @@ where
             mint_context,
         );
 
-        let ret: CLValue = match entry_point_name {
+        let system_config = protocol_data.system_config();
+        let mint_costs = system_config.mint_costs();
+
+        let result = match entry_point_name {
             // Type: `fn mint(amount: U512) -> Result<URef, Error>`
-            mint::METHOD_MINT => {
+            mint::METHOD_MINT => (|| {
+                mint_runtime.charge_system_contract_call(mint_costs.mint)?;
+
                 let amount: U512 = Self::get_named_argument(&runtime_args, mint::ARG_AMOUNT)?;
                 let result: Result<URef, system_contract_errors::mint::Error> =
                     mint_runtime.mint(amount);
                 if let Err(system_contract_errors::mint::Error::GasLimit) = result {
                     return Err(execution::Error::GasLimit);
                 }
-                CLValue::from_t(result)?
-            }
-            mint::METHOD_REDUCE_TOTAL_SUPPLY => {
+                CLValue::from_t(result).map_err(Self::reverter)
+            })(),
```

Comment:
Yes, correct. A workaround for a lack of stable try blocks https://doc.rust-lang.org/beta/unstable-book/language-features/try-blocks.html. There was a crate that implemented `try_block!` macro but it didn't work with mutable lambdas. Once try blocks will be stabilized it would be something like:

```rust
method::FOO => try {
  something()?;
  something()?;
  Ok(())
  },
```